torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -437,6 +463,251 @@ std::vector<at::Tensor> FusionExecutorCache::runFusionWithInputs(
       inputs, launch_params, unique_id);
 }
 
+FusionSegmentRuntime::FusionSegmentRuntime(
+    SegmentedFusion* segmented_fusion,
+    std::unique_ptr<SegmentHeuristics>& heuristics,
+    size_t input_id)
+    : executors_(segmented_fusion->groups().size()),
+      heuristics_(std::move(heuristics)),
+      cache_id_(input_id),
+      segmented_fusion_(segmented_fusion) {}
+
+// Largely duplicated from FusionExecutorCache
+std::vector<at::Tensor> FusionSegmentRuntime::runSegmentWithInput(
+    SegmentedGroup* sg,
+    const at::ArrayRef<IValue>& inputs,
+    size_t input_id) {
+  auto group_id = sg->groupId();
+  const int device_index = getCommonDeviceCUDA(inputs);
+  LaunchParams launch_params;
+
+  auto scheduler_entry = schedulers()[group_id].get();
+
+  // Check that the heuristics are matched
+  TORCH_INTERNAL_ASSERT(scheduler_entry->heuristc() == sg->heuristic());
+
+  if (!executors_[group_id].compiled()) {
+    std::unique_ptr<Fusion> fusion_seg = segmented_fusion_->makeFusion(sg);
+    CompileOptions options;
+    options.device = c10::Device(DeviceType::CUDA, device_index);
+    scheduler_entry->schedule(fusion_seg.get());
+    executors_[group_id].compileFusion(fusion_seg.get(), options);
+  }
+
+  // Load launch params for reduction and normalization kernels
+  if (scheduler_entry->hasParam()) {
+    launch_params = scheduler_entry->params().lparams;
+  }
+
+  return executors_[group_id].runFusion(inputs, launch_params, input_id);
+}
+
+std::vector<at::Tensor> FusionSegmentRuntime::runWithInput(
+    const at::ArrayRef<IValue>& inputs,
+    size_t input_id) {
+  TORCH_INTERNAL_ASSERT(
+      inputs.size() == segmented_fusion_->inputs().size(),
+      "Inputs were not set up correctly, recieved ",
+      inputs.size(),
+      " inputs but expecting ",
+      segmented_fusion_->inputs().size());
+
+  // Map to keep track of currently available tensors
+  std::unordered_map<Val*, IValue> tensor_map;
```

Comment:
nitpick: `*IValue` instead.

Reply:
It'd be good idea for the fusion inputs. I went back and forth on this and decided to still keep it this way. The complication was that `tensor_map` needs to keep the life time of the intermediate kernel output tensors as well and I couldn't quickly figure out a way to nicely re-structure this.